core/src/main/java/jenkins/model/Jenkins.java

```diff
@@ -1033,9 +1033,12 @@ public InstallState getInstallState() {
     public void setInstallState(@Nonnull InstallState newState) {
         InstallState prior = installState;
         installState = newState;
-        if (!prior.equals(newState)) {
+        LOGGER.info("setInstallState (Previous:" + prior + ", New: " + installState + ")");
+        if (!newState.equals(prior)) {
             newState.initializeState();
+            getSetupWizard().onInstallStateUpdate(newState);
         }
+        saveQuietly();
```

Comment:
Is it safe to save the entire Jenkins instance at the current state. IIUC it may be not fully initialized though the state should likely indicate it

Reply:
`SetupWizard` is initialized after `InitMilestone.COMPLETED` so AFAIU this is safe.